src/garage/sampler/rl2_sampler.py

```diff
@@ -0,0 +1,195 @@
+                  :math:`[N, S^*]`. One example is "ale.lives" for atari
+                  environments.
+                * agent_infos: A dictionary with each key representing one
+                  agent info, value being a numpy.ndarray with shape
+                  :math:`[N, S^*]`. One example is "prev_action", which is used
+                  for recurrent policy as previous action input, merged with
+                  the observation input as the state input.
+
+        """
+        logger.log('Obtaining samples for iteration %d...' % itr)
+
+        if batch_size is None:
+            batch_size = (self._episode_per_task * self.algo.max_path_length *
+                          self._n_envs)
+
+        paths = OrderedDict()
+        for i in range(self._n_envs):
+            paths[i] = []
+
+        n_samples = 0
+        obses = self._vec_env.reset()
+        dones = np.asarray([True] * self._vec_env.num_envs)
+        running_paths = [None] * self._vec_env.num_envs
+
+        pbar = ProgBarCounter(batch_size)
+        policy_time = 0
+        env_time = 0
+        process_time = 0
+
+        policy = self.algo.policy
+        # Only reset policies at the beginning of a meta batch
+        policy.reset(dones)
+
+        while n_samples < batch_size:
+            t = time.time()
+
+            actions, agent_infos = policy.get_actions(obses)
+
+            policy_time += time.time() - t
+            t = time.time()
+            next_obses, rewards, dones, env_infos = self._vec_env.step(actions)
+            env_time += time.time() - t
+            t = time.time()
+
+            agent_infos = tensor_utils.split_tensor_dict_list(agent_infos)
+            env_infos = tensor_utils.split_tensor_dict_list(env_infos)
+            if env_infos is None:
+                env_infos = [dict() for _ in range(self._vec_env.num_envs)]
+            if agent_infos is None:
+                agent_infos = [dict() for _ in range(self._vec_env.num_envs)]
+            for idx, observation, action, reward, env_info, agent_info, done in zip(  # noqa: E501
+                    itertools.count(), obses, actions, rewards, env_infos,
+                    agent_infos, dones):
+                if running_paths[idx] is None:
+                    running_paths[idx] = dict(
+                        observations=[],
+                        actions=[],
+                        rewards=[],
+                        dones=[],
+                        env_infos=[],
+                        agent_infos=[],
+                    )
+                running_paths[idx]['observations'].append(observation)
+                running_paths[idx]['actions'].append(action)
+                running_paths[idx]['rewards'].append(reward)
+                running_paths[idx]['dones'].append(done)
+                running_paths[idx]['env_infos'].append(env_info)
+                running_paths[idx]['agent_infos'].append(agent_info)
```

Comment:
I think you mentioned in Slack that additional data (something like `task_id`, IIRC) is needed so that an algorithm can split paths from different tasks. Since you are not adding it here, do you find another workaround?

Reply:
Yes. In order to split the paths, I construct an array by `np.cumsum([len(path['rewards']) for path in paths])` so I can split the paths by `np.split` using this array.

Reply:
Ah. Smart solution.